Sources/Apollo/URLSessionClient.swift

```diff
@@ -0,0 +1,246 @@
+    self.datas.removeValue(forKey: identifier)
+    self.responses.removeValue(forKey: identifier)
+  }
+  
+  /// Clears underlying dictionaries of any data related to all tasks.
+  ///
+  /// Mostly useful for cleanup and/or after invalidation of the `URLSession`.
+  open func clearAllTasks() {
+    self.rawCompletions.removeAll()
+    self.completionBlocks.removeAll()
+    self.datas.removeAll()
+    self.responses.removeAll()
+  }
+  
+  /// The main method to perform a request.
+  ///
+  /// - Parameters:
+  ///   - request: The request to perform.
+  ///   - rawTaskCompletionHandler: [optional] A completion handler to call once the raw task is done, so if an Error requires access to the headers, the user can still access these.
+  ///   - completion: A completion handler to call when the task has either completed successfully or failed.
+  ///
+  /// - Returns: The created URLSesssion task, already resumed, because nobody ever remembers to call `resume()`.
+  @discardableResult
+  open func sendRequest(_ request: URLRequest,
+                        rawTaskCompletionHandler: RawCompletion? = nil,
+                        completion: @escaping Completion) -> URLSessionTask {
+    let dataTask = self.session.dataTask(with: request)
+    if let rawCompletion = rawTaskCompletionHandler {
+      self.rawCompletions[dataTask.taskIdentifier] = rawCompletion
+    }
+    
+    self.completionBlocks[dataTask.taskIdentifier] = completion
+    self.datas[dataTask.taskIdentifier] = Data()
+    dataTask.resume()
+    
+    return dataTask
+  }
+  
+  /// Cancels a given task and clears out its underlying data.
+  ///
+  /// NOTE: You will not receive any kind of "This was cancelled" error when this is called.
+  ///
+  /// - Parameter task: The task you wish to cancel.
+  open func cancel(task: URLSessionTask) {
+    self.clearTask(with: task.taskIdentifier)
+    task.cancel()
+  }
+  
+  // MARK: - URLSessionDelegate
+  
+  open func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?) {
+    let finalError = error ?? URLSessionClientError.sessionBecameInvalidWithoutUnderlyingError
+    for block in completionBlocks.values {
+      block(.failure(finalError))
+    }
+    
+    self.clearAllTasks()
+  }
+  
+  @available(OSX 10.12, iOS 10.0, tvOS 10.0, *)
+  open func urlSession(_ session: URLSession,
+                       task: URLSessionTask,
+                       didFinishCollecting metrics: URLSessionTaskMetrics) {
+    // No default implementation
+  }
+  
+  open func urlSession(_ session: URLSession,
+                       didReceive challenge: URLAuthenticationChallenge,
+                       completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+    completionHandler(.performDefaultHandling, nil)
+  }
+  
+  #if os(iOS) || os(tvOS) || os(watchOS)
+  open func urlSessionDidFinishEvents(forBackgroundURLSession session: URLSession) {
```

Comment:
Shouldn't we kind of warning the developers that they should call the `completionHandler` from `application(_:handleEventsForBackgroundURLSession:completionHandler:)` after this method had been called?
Maybe allow the developer to pass a closure that could be called after the event be triggered?

Something like:
```
open func urlSessionDidFinishEvents(forBackgroundURLSession session: URLSession) {
    self.sessionDidFinishEvents?()
}
```

Reply:
This is why this is `open` - so that it can be subclassed if needed and the developer can handle this as needed.